inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/support/AnnotationQueueOverviewDataProvider.java

```diff
@@ -300,60 +350,111 @@ public long getInProgressAmountForDocument(
 
     //Returns a random document out of all documents in the project.
     //Only a document is chosen which is not yet given to annotators more than the default number
-    //per document number
-    public SourceDocument getRandomDocument()
+    //per document number (TODO),
+    //However, if there was already a document in progress,
+    // this will always be returned (can only be one)
+
+    public SourceDocument getRandomDocument(
+        AnnotationPageBase aAnnotationPageBase, AnnotationDocument aCurrentAnno)
     {
-        //Create an empty document
-        SourceDocument document = null;
+
+        //First check for a documents that is in Progress for the user, and return this
+        for (int i = 0; i < allAnnotationDocuments.size(); i++)
+        {
+            if (allAnnotationDocuments.get(i).getUser().equals
+                (aAnnotationPageBase.getModelObject().getUser().getUsername())
+                && allAnnotationDocuments.get(i).getState().equals(IN_PROGRESS)
+                && !allAnnotationDocuments.get(i).getName().equals(aCurrentAnno.getName()))
+            {
+                //Found a document in progress
+                for (SourceDocument doc: data)
+                {
+                    if (doc.getName().equals(allAnnotationDocuments.get(i).getName()))
+                    {
+                        return doc;
+                    }
+                }
+            }
+        }
+
+
         Random r = new Random();
+        int attempts = 0;
 
-        while (document == null)
+        //Nothing in Progress, so now return a random document, or create a new annotation document
+        while (true)
         {
-            int i = r.nextInt(data.size() - 1);
-            //If the random chosen document wont surpass the amount of default number combining
-            // "inProgress" for the document + "finished" amount for the document + 1
-            if ((getInProgressAmountForDocument(data.
-                get(i)) +
-                getFinishedAmountForDocument(data.
-                    get(i)))
-                + 1 <= defaultAnnotations)
+            SourceDocument src = data.get(r.nextInt(data.size()));
+
+            for (int i = 0; i < allAnnotationDocuments.size(); i++)
             {
-                //If that was not the case, assign this document
-                document = data.get(r.nextInt(data.size() - 1));
+                if (src.getName().equals(allAnnotationDocuments.get(i).getName()) &&
+                    allAnnotationDocuments.get(i).getUser().equals(
+                        aAnnotationPageBase.getModelObject().getUser().getUsername()))
+                {
+                    if (allAnnotationDocuments.get(i).getState().equals(NEW))
+                    {
+                        return src;
+                    } else {
+                        attempts++;
+                        //Try for 3 times the size of all documents, if still nothing new found
```

Comment:
What is going on here? We do you need to attempt stuff? In the optimal case, you could whip up a database query which provides a list of candidate documents and even selects one of them randomly in the DB (cf. https://vladmihalcea.com/sql-order-by-random/ - section on JPA / Hibernate).